### ISSUE TITLE
POSIX::strftime bugfeature: incorrect mktime() call

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#838]
Migrated from [rt.perl.org#1766](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=1766) (status was 'resolved')

Searchable as RT1766$

**Comment by p5pRT** [comment]
### From netch@lucky.net



GMTime of unixtime 941284799 is 30th of October\, 1999\, 11&#8203;:59&#8203;:59\.

The following code&#8203;:

\#\!/usr/bin/perl
use POSIX;
print strftime\("%Y %m %d %H %M %S"\, gmtime\(941284799\)\)\, "\\n";

prints

1999 10 30 12 59 59

\(note environment information&#8203;: timezone is Europe/Kiev\)

This effect \(adding of 1 hour\) appeared in perl 5\.00503 and did not exist in
perl 5\.00502\.

Diff of $\{perl\}/ext/POSIX/POSIX\.xs between 5\.00502 and 5\.00503
contains following&#8203;:

===\{
@&#8203;@&#8203; \-3591\,7 \+3603\,7 @&#8203;@&#8203;
&nbsp;        RETVAL

 char \*
\-strftime\(fmt\, sec\, min\, hour\, mday\, mon\, year\, wday = 0\, yday = 0\, isdst = 0\)
\+strftime\(fmt\, sec\, min\, hour\, mday\, mon\, year\, wday = \-1\, yday = \-1\, isdst = \-1
&nbsp;        char \*          fmt
&nbsp;        int             sec
&nbsp;        int             min
@&#8203;@&#8203; \-3617\,8 \+3629\,45 @&#8203;@&#8203;
&nbsp;            mytm\.tm\_wday = wday;
&nbsp;            mytm\.tm\_yday = yday;
&nbsp;            mytm\.tm\_isdst = isdst;
\+           \(void\) mktime\(&mytm\);
&nbsp;            len = strftime\(tmpbuf\, sizeof tmpbuf\, fmt\, &mytm\);
===\}

Well\, test it and see that mktime\(\) normalizes time according to
local time zone \(Europe/Kiev in our case\)&#8203;:

===\{
\#include \<time\.h>
\#include \<stdio\.h>
\#include \<string\.h>
\#include \<errno\.h>

void f\( int x \)
\{
&nbsp;   char buf\[ 200 \];
&nbsp;   struct tm stm;
&nbsp;   bzero\( &stm\, sizeof stm \);
&nbsp;   stm\.tm\_year = 99; stm\.tm\_mon = 9; stm\.tm\_mday = 30; stm\.tm\_hour = 11;
&nbsp;   stm\.tm\_min = 59; stm\.tm\_sec = 59; stm\.tm\_isdst = 0;
&nbsp;   if\( x \)
&nbsp;      mktime\( &stm \);
&nbsp;   bzero\( buf\, sizeof buf \);
&nbsp;   strftime\( buf\, sizeof buf\, "%Y %m %d %H %M %S"\, &stm \);
&nbsp;   puts\( buf \);
&nbsp;   printf\( "%d %d %d %d %d %d\\n"\, stm\.tm\_year\, stm\.tm\_mon\, stm\.tm\_mday\,
&nbsp;         stm\.tm\_hour\, stm\.tm\_min\, stm\.tm\_sec\, stm\.tm\_isdst \);
\}

int main\(\) \{
&nbsp;   f\( 0 \); f\( 1 \); return 1;
\}
===\}

output is&#8203;:

===\{
netch@&#8203;burka&#8203;:\~/prog/tiny/2>\./3
1999 10 30 11 59 59
99 9 30 11 59 59 0
1999 10 30 12 59 59
99 9 30 12 59 59 1
===\}

This mktime\(\) behavior possibly correct and in any case accords to its man page&#8203;:

===\{
&nbsp;     On successful completion\, the values of the tm\_wday and tm\_yday compo\-
&nbsp;     nents of the structure are set appropriately\, and the other components
&nbsp;     are set to represent the specified calendar time\, but with their values
&nbsp;     forced to their normal ranges; the final value of tm\_mday is not set un\-
&nbsp;     til tm\_mon and tm\_year are determined\.  Mktime\(\) returns the specified
&nbsp;     calendar time; if the calendar time cannot be represented\, it returns \-1;
===\}

But\, the time in question was GMT time\, not local time\.

Thus\, perl MUST NOT call mktime\(\) in strftime\(\) because strftime MUST
ONLY PRINT its data and MUST NOT have any opinion of its content because
it cannot know real time zone of the data\.


&nbsp;  >Fix&#8203;: 
&nbsp;    
Disable the mktime\(\) call in POSIX&#8203;::strftime\.

\(For FreeBSD only\. This text originally was PR to FreeBSD\.\)
Also disable init\_tm\(\)\,
whis is really localtime\(time\(\)\) \- IMHO the better solution for FreeBSD
in case of tm\_gmtoff & tm\_zone patameters is to set them to most safe value\,
i\.e\. 0\.

\(note&#8203;: spaces/tabs are incorrect in this diff\)
===\{
<details><summary>Inline Patch</summary>

```diff
--- src/contrib/perl5/ext/POSIX/POSIX.xs.orig   Wed May  5 16:15:29 1999
+++ src/contrib/perl5/ext/POSIX/POSIX.xs        Wed Nov 10 20:04:30 1999
@@ -3617,21 +3617,20 @@
     CODE:
        {
            char tmpbuf[128];
            struct tm mytm;
            int len;
-           init_tm(&mytm);     /* XXX workaround - see init_tm() above */
+           bzero(&mytm, sizeof(mytm));
            mytm.tm_sec = sec;
            mytm.tm_min = min;
            mytm.tm_hour = hour;
            mytm.tm_mday = mday;
            mytm.tm_mon = mon;
            mytm.tm_year = year;
            mytm.tm_wday = wday;
```
</details>

&nbsp;            mytm\.tm\_yday = yday;
&nbsp;            mytm\.tm\_isdst = isdst;
\-           \(void\) mktime\(&mytm\);
&nbsp;            len = strftime\(tmpbuf\, sizeof tmpbuf\, fmt\, &mytm\);
&nbsp;            /\*
&nbsp;            \*\* The following is needed to handle to the situation where
&nbsp;            \*\* tmpbuf overflows\.  Basically we want to allocate a buffer
&nbsp;            \*\* and try repeatedly\.  The reason why it is so complicated
===\}

<details><summary>Perl Info</summary>

```


Site configuration information for perl 5.00503:

Configured by markm at $Date: 1999/05/05 19:42:40 $.

Summary of my perl5 (5.0 patchlevel 5 subversion 3) configuration:
  Platform:
    osname=freebsd, osvers=4.0-current, archname=i386-freebsd
    uname='freebsd freefall.freebsd.org 4.0-current freebsd 4.0-current #0: $Date: 1999/05/05 19:42:40 $'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef useperlio=undef d_sfio=undef
  Compiler:
    cc='cc', optimize='undef', gccversion=egcs-2.91.66 19990314 (egcs-1.1.2 release)
    cppflags=''
    ccflags =''
    stdchar='char', d_stdstdio=undef, usevfork=true
    intsize=4, longsize=4, ptrsize=4, doublesize=8
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags ='-Wl,-E'
    libpth=/usr/lib
    libs=-lm -lc -lcrypt
    libc=/usr/lib/libc.so, so=so, useshrplib=true, libperl=libperl.so.3
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-DPIC -fpic', lddlflags='-shared'

Locally applied patches:
    


@INC for perl 5.00503:
    /usr/libdata/perl/5.00503/mach
    /usr/libdata/perl/5.00503
    /usr/local/lib/perl5/site_perl/5.005/i386-freebsd
    /usr/local/lib/perl5/site_perl/5.005
    .


Environment for perl 5.00503:
    HOME=/usr/homes/netch
    LANG=ru_RU.KOI8-R
    LANGUAGE (unset)
    LC_COLLATE=ru_RU.KOI8-R
    LC_CTYPE=ru_RU.KOI8-R
    LC_TIME=C
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/homes/netch/bin:/usr/local/bin:/usr/local/sbin:/usr/bin:/usr/sbin:/bin:/sbin
    PERL_BADLANG (unset)
    SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Wed\, 10 Nov 1999 at 21&#8203;:51&#8203;:22 \+0200\, Valentin Nechayev wrote&#8203;:

> This effect \(adding of 1 hour\) appeared in perl 5\.00503 and did not exist in
> perl 5\.00502\.

This is fixed\, it appears\, in 5\.00562 \[\*\]

 \[  4223\] By&#8203;: gsar                                  on 1999/09/24 05&#8203;:05&#8203;:06
&nbsp;         Log&#8203;: normalize time for strftime\(\) \(without the isdst effects of
&nbsp;              mktime\(\)\) using a custom mini\_mktime\(\)
&nbsp;              From&#8203;: spider\-perl@&#8203;Orb\.Nashua\.NH\.US
&nbsp;              Date&#8203;: Thu\, 23 Sep 1999 17&#8203;:54&#8203;:53 \-0400
&nbsp;              Message\-Id&#8203;: \<199909232154\.RAA25151@&#8203;leggy\.zk3\.dec\.com>
&nbsp;              Subject&#8203;: Re&#8203;: \[ID 19990913\.003\] Possible bug using POSIX&#8203;::strftime

One cannot do without this normalisation if the optionality of the three
trailing parameters is to be preserved\. The versions of strftime before
5\.00503 got things totally wrong\, or core\-dumped\, because of this\.

Ian

\[\*\] I really must get round to loading 5\.00562\.\.\. unfortunately\, my
&nbsp;    patch \(2\.1b\) barfs on the 5\.00562 patch file \- too long a line in
&nbsp;    the big\-float test module\.\.\.&#8203;:\)
